zebra-state/src/sled_state.rs

```diff
@@ -290,35 +220,69 @@ impl FinalizedState {
             &self.height_by_hash,
             &self.block_by_height,
             &self.utxo_by_outpoint,
+            &self.tx_by_hash,
+            &self.sprout_nullifiers,
+            &self.sapling_nullifiers,
         )
             .transaction(
-                move |(hash_by_height, height_by_hash, block_by_height, utxo_by_outpoint)| {
-                    // TODO: do serialization above
-                    // for some reason this wouldn't move into the closure (??)
-                    let block_bytes = block
-                        .zcash_serialize_to_vec()
-                        .expect("zcash_serialize_to_vec has wrong return type");
-
+                move |(
+                    hash_by_height,
+                    height_by_hash,
+                    block_by_height,
+                    utxo_by_outpoint,
+                    tx_by_hash,
+                    sprout_nullifiers,
+                    sapling_nullifiers,
+                )| {
                     // TODO: check highest entry of hash_by_height as in RFC
 
-                    hash_by_height.insert(&height_bytes, &hash.0)?;
-                    height_by_hash.insert(&hash.0, &height_bytes)?;
-                    block_by_height.insert(&height_bytes, block_bytes)?;
-                    // tx_by_hash
+                    // Index the block
+                    hash_by_height.zs_insert(height, hash)?;
+                    height_by_hash.zs_insert(hash, height)?;
+                    block_by_height.zs_insert(height, &*block)?;
+
+                    // TODO: sprout and sapling anchors (per block)
 
+                    // Consensus-critical bug in zcashd: transactions in the
+                    // genesis block are ignored.
+                    if block.header.previous_block_hash == block::Hash([0; 32]) {
+                        return Ok(hash);
+                    }
```

Comment:
I think the sprout anchor from the genesis block is also ignored, but let's confirm that.

```suggestion
                    // Consensus-critical bug in zcashd: transactions in the
                    // genesis block are ignored.
                    if block.header.previous_block_hash == block::Hash([0; 32]) {
                        return Ok(hash);
                    }
                    
                    // TODO: sprout and sapling anchors (per block)
```

Reply:
@dconnolly do we only need to store one sprout anchor per block?
Or is every sprout treestate, after every transaction, a valid sprout anchor for future transactions?

Reply:
Hypothetically, if there were Sapling anchors at genesis, `zcashd` would ignore them.

I don't know what Zebra should do with both the sprout and sapling edge cases, see PR #1232 for a potential RFC update.